src/internal/operators/takeWhile.ts

```diff
@@ -3,10 +3,10 @@ import { Observable } from '../Observable';
 import { Subscriber } from '../Subscriber';
 import { OperatorFunction, MonoTypeOperatorFunction, TeardownLogic } from '../types';
 
-export function takeWhile<T, S extends T>(predicate: (value: T, index: number) => value is S): OperatorFunction<T, S>;
-export function takeWhile<T, S extends T>(predicate: (value: T, index: number) => value is S, inclusive: false): OperatorFunction<T, S>;
-export function takeWhile<T>(predicate: (value: T, index: number) => boolean, inclusive?: boolean): MonoTypeOperatorFunction<T>;
-
+export function takeWhile<T>(predicate: BooleanConstructor, inclusive?: boolean): OperatorFunction<T, T>;
+export function takeWhile<T, S extends T>(predicate: (value: T, index: number) => value is S, inclusive: true): OperatorFunction<T, T>;
+export function takeWhile<T, S extends T>(predicate: (value: T, index: number) => value is S, inclusive?: boolean): OperatorFunction<T, S>;
```

Comment:
I think this is the wrong way around. If a boolean variable is passed - instead of a literal - it will match the second of the highlighted signatures and the return type will be `OperatorFunction<T, S>`, but if the variable is `true` at runtime, a non-`S` type could be emitted.

I think that a user-defined type guard can be used as the predicate **only** when `inclusive` is known to be `false` at compile time. So the signature with the user-defined type guard should be:

```
export function takeWhile<T, S extends T>(predicate: (value: T, index: number) => value is S, inclusive: false): OperatorFunction<T, S>;
```

And the other should use a garden-variety predicate and not a user-defined type guard.